### PR TITLE
Merge Sindarin BCOffsetCache

### PR DESCRIPTION
**Pull request by MarcusDenker**
This PR integrates the SindarinBytecodeToASTCache as OCBytecodeToASTCache

=> API changed to require a RBMethodNode, not the compiledMethod
=> improved the case when asking for start pc for methods with primitives

This is another step towards making the IR an transient datastructure that is not cached

### CODE REVIEW

File: src/OpalCompiler-Core/OCBytecodeToASTCache.class.st [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+{ #category : #initialization }
+OCBytecodeToASTCache >> generateForMethodNode: aNode [
+	| methodIR currentBcOffset |
+	methodNode := aNode.
+	methodIR := methodNode ir.
+	bcToASTMap := Dictionary new.
+	firstBcOffset := methodNode compiledMethod initialPC.
+	
+	"if there is a primitive, we need to record the offset"
+	(methodIR irPrimitive num ~= 0) ifTrue: [ bcToASTMap at: firstBcOffset put: methodNode].
```
Comment (StevenCostiou) [inline src/OpalCompiler-Core/OCBytecodeToASTCache.class.st:68]: This case with primitives is not covered by a test. 
Could you add one?

Could you give more details about that case? I would like to document it in the compiler booklet.

File: src/OpalCompiler-Core/RBMethodNode.extension.st [hunk 2]
```diff
@@ -167,7 +167,8 @@ RBMethodNode >> rewriteTempsForContext: aContext [
 
 { #category : #'*OpalCompiler-Core' }
 RBMethodNode >> sourceNodeForPC: anInteger [
-	^(self ir instructionForPC: anInteger) sourceNode
+	bcToASTCache ifNil: [ bcToASTCache := OCBytecodeToASTCache generateForMethodNode: self].
```
Comment (StevenCostiou) [inline src/OpalCompiler-Core/RBMethodNode.extension.st]: If we add a lazy accessor to bcToASTCache, we could just have 
`^ self bcToASTCache nodeForPC: anInteger`.

What do you think?

**Review by StevenCostiou — CHANGES_REQUESTED** [review]

File: src/OpalCompiler-Core/OCBytecodeToASTCache.class.st (hunk 1 above)
Comment (MarcusDenker) [reply src/OpalCompiler-Core/OCBytecodeToASTCache.class.st:68]: There is #testPrimitiveMethodSourceNodeAtInitialPC which failed without this.

So the thing is that primitives are modelled with a bytecode, but there irPrimitive is not part of the other IR instructions, but instead the IRMethod has an ivar that references is.

This is something we could actually change.

Comment (MarcusDenker) [reply src/OpalCompiler-Core/OCBytecodeToASTCache.class.st:68]: (but not as part of this PR)

File: src/OpalCompiler-Core/RBMethodNode.extension.st (hunk 2 above)
Comment (MarcusDenker) [reply src/OpalCompiler-Core/RBMethodNode.extension.st]: Yes!

**Review by StevenCostiou — APPROVED** [review]

**Pull request merged**